File: src/edgepy/edgepy_cpp/glm_one_group.cpp
// Copyright (C) 2008-2022 Yunshun Chen, Aaron TL Lun, Davis J McCarthy, Matthew E Ritchie, Belinda Phipson, Yifang Hu, Xiaobei Zhou, Mark D Robinson, Gordon K Smyth
// Copyright (C) 2022-2023 Maximilien Colange
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <https://www.gnu.org/licenses/>.

// This file is based on the file 'src/glm_one_group.cpp' of the Bioconductor edgeR package (version 3.38.4).


#include "glm.h"
#include "utils.h"

std::pair<double,bool> glm_one_group(size_t nlibs, const double* counts, const double* offset,
        const double* disp, const double* weights, long maxit, double tolerance, double cur_beta) {
    /* Setting up initial values for beta as the log of the mean of the ratio of counts to offsets.
 	 * This is the exact solution for the gamma distribution (which is the limit of the NB as
 	 * the dispersion goes to infinity. However, if cur_beta is not NA, then we assume it's good.
 	 */
	bool nonzero=false;
	if (isnan(cur_beta)) {
		cur_beta=0;
 	   	double totweight=0;
		for (size_t j=0; j<nlibs; ++j) {
			const double& cur_val=counts[j];
			if (cur_val>low_value) {
				cur_beta+=cur_val/std::exp(offset[j]) * weights[j];
				nonzero=true;
			}
			totweight+=weights[j];
		}
		cur_beta=std::log(cur_beta/totweight);
	} else {
		for (size_t j=0; j<nlibs; ++j) {
			if (counts[j] > low_value) {
                nonzero=true;
                break;
            }
		}
	}

	// Skipping to a result for all-zero rows.
	if (!nonzero) {
        return std::make_pair(neg_inf, true);
    }

	// Newton-Raphson iterations to converge to mean.
    bool has_converged=false;
	for (long i=0; i<maxit; ++i) {
		double dl=0;
 	    double info=0;
		for (size_t j=0; j<nlibs; ++j) {
			const double mu=std::exp(cur_beta+offset[j]), denominator=1+mu*disp[j];
			dl+=(counts[j]-mu)/denominator * weights[j];
			info+=mu/denominator * weights[j];
		}
		const double step=dl/info;
		cur_beta+=step;
		if (std::abs(step)<tolerance) {
			has_converged=true;
			break;
		}
	}

	return std::make_pair(cur_beta, has_converged);
}

File: inmoose/edgepy/glm_one_group.py
# -----------------------------------------------------------------------------
# Copyright (C) 2008-2022 Yunshun Chen, Aaron TL Lun, Davis J McCarthy, Matthew E Ritchie, Belinda Phipson, Yifang Hu, Xiaobei Zhou, Mark D Robinson, Gordon K Smyth
# Copyright (C) 2022-2023 Maximilien Colange

# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.

# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.
# -----------------------------------------------------------------------------

# This file is a Python port of the original C++ code from the files
# 'src/glm_one_group.cpp' and 'src/R_fit_one_group.cpp' of the Bioconductor
# edgeR package (version 3.38.4).

import numpy as np

from .edgepy_cpp import glm_one_group_cython


def glm_one_group(counts, offset, disp, weights, maxit, tolerance, cur_beta):
    """
    Simplified fit for negative binomial GLM when the design matrix is one group

    This is the low-level function called by :func:`fit_one_group`.

    See Also
    --------
    fit_one_group

    Arguments
    ---------
    counts : array_like
        matrix of counts
    offset : array_like
        matrix of offsets for the log-linear models
    disp : array_like
        matrix of dispersions
    weights : array_like
        matrix of observation weights
    maxit : int
        maximum number of Newton-Raphson iterations
    tol : float
        tolerance for convergence in the Newton-Raphson iteration
    cur_beta : array_like
        initial coefficients

    Returns
    -------
    ndarray
        fitted coefficients (one per row in :code:`y`)
    ndarray
        whether the fit converged for each coefficient
    """

    out_beta = cur_beta.copy()
    out_conv = np.repeat(False, out_beta.shape)

    for j in range(out_beta.shape[0]):
        (out_beta[j], out_conv[j]) = glm_one_group_cython(
            counts.shape[1],
            counts[j, :],
            offset[j, :],
            disp[j, :],
            weights[j, :],
            maxit,
            tolerance,
            cur_beta[j],
        )
    return (out_beta, out_conv)


def fit_one_group(y, offsets, disp, weights, maxit, tol, beta, usePoisson=True):
    """
    Simplified fit for negative binomial GLM when the design matrix is one group

    In essence, this function has the same purpose as :func:`fit_levenberg`,
    specialized for the case where the design matrix is one group. In
    particular, Levenberg damping is not required here.

    This function also uses a Poisson approximation for zero dispersion genes,
    and calls :func:`glm_one_group` otherwise.

    See Also
    --------
    fit_levenberg
    glm_one_group

    Arguments
    ---------
    y : array_like
        matrix of counts
    offsets : array_like
        matrix of offsets for the log-linear models
    disp : array_like
        matrix of dispersions
    weights : array_like
        matrix of observation weights
    maxit : int
        maximum number of Newton-Raphson iterations
    tol : float
        tolerance for convergence in the Newton-Raphson iteration
    beta : array_like
        initial coefficients
    usePoisson : bool, optional
        whether to use Poisson approximation for zero-disp one-weight genes.
        Defaults to :code:`True`

    Returns
    -------
    ndarray
        fitted coefficients (one per row in :code:`y`)
    ndarray
        whether the fit converged for each coefficient
    """

    assert beta.shape == (y.shape[0],)

    out_beta = beta.copy()
    out_conv = np.repeat(False, beta.shape)

    # Checking for the Poisson special case with all-unity weights and all-zero
    # dispersions
    if usePoisson:
        disp_is_zero = (disp == 0).all(axis=1)
        weight_is_one = (weights == 1).all(axis=1)
        special_case = disp_is_zero & weight_is_one
        general_case = ~special_case
    else:
        special_case = np.repeat(False, y.shape[0])
        general_case = ~special_case

    sum_lib = np.exp(offsets[special_case]).sum(axis=1)
    sum_counts = y[special_case].sum(axis=1)
    with np.errstate(divide="ignore"):
        out_beta[special_case] = np.where(
            sum_counts == 0, -np.inf, np.log(sum_counts / sum_lib)
        )
    out_conv[special_case] = True

    # Otherwise going through NR iterations
    out_beta[general_case], out_conv[general_case] = glm_one_group(
        y[general_case],
        offsets[general_case],
        disp[general_case],
        weights[general_case],
        maxit,
        tol,
        out_beta[general_case],
    )

    return (out_beta, out_conv)
